modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java

```diff
@@ -82,57 +86,64 @@ protected AbstractIgniteJoinConverterRule(String descriptionPrefix, HintDefiniti
 
     /** */
     private boolean disabledByHints(LogicalJoin join) {
-        if (HintUtils.allRelHints(join).isEmpty())
+        Collection<TableScan> tables = joinTables(join);
+
+        Collection<RelHint> rawHints = new ArrayList<>();
+
+        // Table hints have a bigger priority and go first.
+        tables.forEach(t -> rawHints.addAll(HintUtils.nonInheritedRelHints(t)));
+
+        rawHints.addAll(HintUtils.allRelHints(join));
+
+        if (rawHints.isEmpty())
             return false;
 
         boolean ruleDisabled = false;
 
         Map<String, Collection<HintDefinition>> hintedTables = new HashMap<>();
 
-        Set<String> joinTbls = joinTblNames(join);
-
-        assert joinTbls.size() < 3;
+        Set<String> tblNames = extractTblNames(tables);
 
-        for (RelHint hint : HintUtils.hints(join, ALL_HINTS)) {
-            Set<String> matchedTbls = hint.listOptions.isEmpty() ? joinTbls : new HashSet<>(hint.listOptions);
+        for (RelHint hint : HintUtils.hints(join, rawHints, ALL_HINTS)) {
+            Set<String> matchedTbls = hint.listOptions.isEmpty() ? tblNames : new HashSet<>(hint.listOptions);
 
             if (!hint.listOptions.isEmpty())
-                matchedTbls.retainAll(joinTbls);
+                matchedTbls.retainAll(tblNames);
 
+            // Do not skip if the hint has no option. It can be a 'global', request-level hint.
             if (matchedTbls.isEmpty())
```

Comment:
Looks like we can change condition to `matchedTbls.isEmpty() && !joinTbls.isEmpty()` with the same effect, and it will be simplier than introduced approach with UNALLOWED_TABLE_NAME